lib/ansible/plugins/strategy/__init__.py

```diff
@@ -481,6 +481,8 @@ def parent_handler_match(target_handler, handler_name):
                                 if original_task.action == 'set_fact':
                                     self._variable_manager.set_nonpersistent_facts(target_host, result_item['ansible_facts'].copy())
                                 else:
+                                    if original_task.action == 'setup':
+                                        target_host._gathered_facts = True
```

Comment:
A 'module_setup' fact sounds like a good indicator. Would it make sense to set _gathered_facts here based on that fact, or try to check 'module_setup' fact directly in play_interator?